### PR TITLE
MARKET-1587 Animate on scroll

### PR DESCRIPTION
**Pull request by StasivPavlo**

### CODE REVIEW

File: components/bl-AnimateOnScroll-component/README.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+|----------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------|---------------|--------------|------------|-------------------------------------------------------------------------------------------------------------------|
+| Disable        | Select ["mobile", "tablet", "true", "false"]                                                                                                                                                                                                                                                                                                   | "false"       | Disable Logic | YES          | YES        | Allows to determine whether or not to disable the animation                                                       |
```
Comment (addKiwi) [inline components/bl-AnimateOnScroll-component/README.md:11]:
```suggestion
| Disable        | *Select* <br/>[`mobile`, `tablet`, `true`, `false`]                                                                                                                                                                                                                                                                                                   | "false"       | Disable Logic | YES          | YES        | Allows to determine whether or not to disable the animation                                                       |
```
We make select in this way [Example](https://github.com/Backendless/ui-builder-library/blob/bl-AnimateOnScroll-component/components/bl-leafletMap-component/README.md#properties)

Comment (addKiwi) [reply components/bl-AnimateOnScroll-component/README.md:11]: Also all types are in italic style ` *Number* `

File: components/bl-AnimateOnScroll-component/component.json [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+      "hasLogicHandler": true,
+      "handlerId": "anchorLogic",
+      "dataBinding": true,
+      "handlerDescription": "This handler allows determining the element by class whose offset will be used to trigger animation instead of an actual one.",
+      "handlerLabel": "Anchor Logic"
+    },
+    {
+      "type": "number",
+      "name": "offset",
+      "label": "Offset",
```
Comment (v-excelsior) [inline components/bl-AnimateOnScroll-component/component.json:223]: Could we add settingTooltip?

File: components/bl-AnimateOnScroll-component/component.json [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+      "type": "select",
+      "name": "disable",
+      "label": "Disable",
+      "showInSettings": true,
+      "defaultValue": "false",
+      "hasLogicHandler": true,
+      "handlerId": "disableLogic",
+      "handlerLabel": "Disable Logic",
+      "dataBinding": true,
+      "options": [
```
Comment (v-excelsior) [inline components/bl-AnimateOnScroll-component/component.json:21]: Do we handle case when user returns boolean true from this handler(as it works in all other components) ?

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-AnimateOnScroll-component/src/index.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    document.addEventListener('aos:in', () => {
+      animatedIn();
+    });
+
+    document.addEventListener('aos:out', () => {
+      animatedOut();
+    });
+  }, []);
+
```
Comment (v-excelsior) [inline components/bl-AnimateOnScroll-component/src/index.js:34]: need to add component.el property with a ref to the root div

File: components/bl-AnimateOnScroll-component/src/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    document.addEventListener('aos:in', () => onAnimationEnter());
+    document.addEventListener('aos:out', () => onAnimationOut());
+
+    return () => {
+      document.removeEventListener('aos:in', () => onAnimationEnter());
+      document.removeEventListener('aos:out', () => onAnimationOut());
+    }
+  }, []);
+
+  component.el = animationOnScrollRef.current;
```
Comment (Valodya) [inline components/bl-AnimateOnScroll-component/src/index.js:39]: I'm not sure this is the correct way

let's imaging this was rendered only once and the `animationOnScrollRef.current` before first render will be null and as result it won't work 

I assume we need to wrap it with `useEffect`

File: components/bl-AnimateOnScroll-component/src/index.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  const elementPod = pods['animationPod'];
+
+  const animationOnScrollRef = useRef(null);
+
+  const animationName = useMemo(() => {
+    return animationType + (side === 'none' ? '' : `-${ side }`);
+  }, [animationType, side]);
+
+  useEffect(() => {
+    Aos.init({ disable: DevicesMap[disableFor] });
```
Comment (Valodya) [inline components/bl-AnimateOnScroll-component/src/index.js:28]: will it work if we modify the `disableFor ` property in runtime?

Comment (StasivPavlo) [reply components/bl-AnimateOnScroll-component/src/index.js:28]: no, that's why I deleted the logic handler and data binding

File: components/bl-AnimateOnScroll-component/src/index.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+  useEffect(() => component.el = animationOnScrollRef.current, []);
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <div
+      ref={ animationOnScrollRef }
```
Comment (Valodya) [inline components/bl-AnimateOnScroll-component/src/index.js:47]: `elRef`

File: components/bl-AnimateOnScroll-component/src/index.js (hunk 6 above)
Comment (Valodya) [reply components/bl-AnimateOnScroll-component/src/index.js:28]: can we specify the `disable` in HTML as `data-aos-disable`?

File: components/bl-AnimateOnScroll-component/src/index.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+const { cn } = BackendlessUI.CSSUtils;
+
+const DevicesMap = {
+  mobile: 'mobile',
+  tablet: 'tablet',
+  all   : true,
+  none  : false,
+};
+
+export default function AnimateOnScroll({ component, eventHandlers, pods }) {
```
Comment (Valodya) [inline components/bl-AnimateOnScroll-component/src/index.js:13]: have you tried to use more than one instance on the same page with different settings?

File: components/bl-AnimateOnScroll-component/src/index.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  const animationOnScrollRef = useRef(null);
+
+  const animationName = useMemo(() => {
+    return animationType + (side === 'none' ? '' : `-${ side }`);
+  }, [animationType, side]);
+
+  useEffect(() => {
+    Aos.init({ disable: DevicesMap[disableFor] });
+
+    document.addEventListener('aos:in', () => onAnimationEnter());
```
Comment (Valodya) [inline components/bl-AnimateOnScroll-component/src/index.js:30]: how does it work when on the page more than one instance?

File: components/bl-AnimateOnScroll-component/README.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+|----------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------|-------------------|--------------|------------|-------------------------------------------------------------------------------------------------------------------|
+| Disable For    | *Select* <br/>[`mobile`, `tablet`, `all`, `none`]                                                                                                                                                                                                                                                                                                     | "false"       | Disable For Logic | YES          | YES        | Allows to determine whether or not to disable the animation                                                       |
```
Comment (Valodya) [inline components/bl-AnimateOnScroll-component/README.md:11]: `false` is the default value for `Disable For`? but in the available list there is only: [`mobile`, `tablet`, `all`, `none`]

File: components/bl-AnimateOnScroll-component/src/index.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+      className={ cn('bl-customComponent-animateOnScroll', classList) }
+      style={ style }
+      data-aos={ animationName }
+      data-aos-easing={ easing }
+      data-aos-duration={ duration }
+      data-aos-delay={ delay }
+      data-aos-offset={ offset }
+      data-aos-mirror={ mirror }
+      data-aos-once={ once }
+      data-aos-anchor={ anchor ? '.' + anchor : null }
```
Comment (Valodya) [inline components/bl-AnimateOnScroll-component/src/index.js:57]: what does `anchor` look like? is it a className?

File: components/bl-AnimateOnScroll-component/src/index.js (hunk 8 above)
Comment (StasivPavlo) [reply components/bl-AnimateOnScroll-component/src/index.js:13]: yes, all work well

File: components/bl-AnimateOnScroll-component/src/index.js (hunk 6 above)
Comment (StasivPavlo) [reply components/bl-AnimateOnScroll-component/src/index.js:28]: no

File: components/bl-AnimateOnScroll-component/src/index.js (hunk 9 above)
Comment (StasivPavlo) [reply components/bl-AnimateOnScroll-component/src/index.js:30]: AOS dispatches two events on document: aos:in and aos:out whenever any element animates in or out

But I see the possibility of making events for the current element and not for any

File: components/bl-AnimateOnScroll-component/src/index.js (hunk 11 above)
Comment (StasivPavlo) [reply components/bl-AnimateOnScroll-component/src/index.js:57]: yes

**Pull request merged**